docs/modules/ROOT/pages/upgrades.adoc

```diff
@@ -49,35 +49,53 @@ NOTE: We will be using the following module to implement the {i_upgradeable} int
 ----
 #[starknet::contract]
 mod UpgradeableContract {
-    use openzeppelin::access::ownable::Ownable;
-    use openzeppelin::upgrades::Upgradeable;
+    use openzeppelin::access::ownable::Ownable as ownable_component;
+    use openzeppelin::upgrades::Upgradeable as upgradeable_component;
     use openzeppelin::upgrades::interface::IUpgradeable;
     use starknet::ClassHash;
     use starknet::ContractAddress;
 
+    component!(path: ownable_component, storage: ownable, event: OwnableEvent);
+    component!(path: upgradeable_component, storage: upgradeable, event: UpgradeableEvent);
+
+    /// Ownable
+    #[abi(embed_v0)]
+    impl OwnableImpl = ownable_component::OwnableImpl<ContractState>;
+    impl OwnableInternalImpl = ownable_component::InternalImpl<ContractState>;
+
+    /// Upgradeable
+    impl UpgradeableInternalImpl = upgradeable_component::InternalImpl<ContractState>;
+
     #[storage]
-    struct Storage {}
+    struct Storage {
+        #[substorage(v0)]
+        ownable: ownable_component::Storage,
+        #[substorage(v0)]
+        upgradeable: upgradeable_component::Storage
+    }
+
+    #[event]
+    #[derive(Drop, starknet::Event)]
+    enum Event {
+        OwnableEvent: ownable_component::Event,
+        UpgradeableEvent: upgradeable_component::Event
```

Comment:
should we be flattening events?

Reply:
My impression was that we need this only for ERC20 and ERC721 (backward compatibility with indexers). Not sure whether we should flatten all events.

Reply:
why would we want nested events? i tonight not flattening was just to allow "overriding" events, not to expose them nested. that'd be inconsistent, and make tooling/interop more complex

Reply:
any pre existing event should be flattened for backwards compat, this included far more than erc20 and 721, also pausable, ownable, etc

Reply:
> why would we want nested events? i tonight not flattening was just to allow "overriding" events, not to expose them nested. that'd be inconsistent, and make tooling/interop more complex

I think the purpose is to avoid potential clashes among components (the design from the cairo team)

Reply:
> any pre existing event should be flattened for backwards compat, this includes far more than erc20 and 721, also pausable, ownable, etc

That's a fair point, but I still think tokens are probably more important than pausable on this matter

Reply:
> I think the purpose is to avoid potential clashes among components

that makes sense when auto embedding, as a default mechanism. but it's an app-level decision whether to prefix events or not, and there's no standard for it. if anything, standards assume they're flattened.

> That's a fair point, but I still think tokens are probably more important than pausable on this matter

what do you mean by "more important"? not sure what importance implies in this context, but I think we should be compatible and standard (although there's no written standard in Cairoland, there's Ethereum standards like Ownable and defacto standards such as Pausable) independently of importance i.e. even if ERC721 is "more important" than ERC20, we want both to be standard and compatible with existing tooling.

that's for existing standards, we yet need to decide what to do with new ones. this requires a bit more of standard discussion which might exceed this library's implementations.